clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerPartitionAssignor.java

```diff
@@ -0,0 +1,237 @@
+     * to be the same as the class name specified in {@link ConsumerConfig#PARTITION_ASSIGNMENT_STRATEGY_CONFIG}
+     * @return non-null unique name
+     */
+    String name();
+
+    final class Subscription {
+        private final List<String> topics;
+        private final ByteBuffer userData;
+        private final List<TopicPartition> ownedPartitions;
+        private Optional<String> groupInstanceId;
+
+        public Subscription(List<String> topics, ByteBuffer userData, List<TopicPartition> ownedPartitions) {
+            this.topics = topics;
+            this.userData = userData;
+            this.ownedPartitions = ownedPartitions;
+            this.groupInstanceId = Optional.empty();
+        }
+
+        public Subscription(List<String> topics, ByteBuffer userData) {
+            this(topics, userData, Collections.emptyList());
+        }
+
+        public Subscription(List<String> topics) {
+            this(topics, null, Collections.emptyList());
+        }
+
+        public List<String> topics() {
+            return topics;
+        }
+
+        public ByteBuffer userData() {
+            return userData;
+        }
+
+        public List<TopicPartition> ownedPartitions() {
+            return ownedPartitions;
+        }
+
+        public void setGroupInstanceId(Optional<String> groupInstanceId) {
+            this.groupInstanceId = groupInstanceId;
+        }
+
+        public Optional<String> groupInstanceId() {
+            return groupInstanceId;
+        }
+    }
+
+    final class Assignment {
+        private List<TopicPartition> partitions;
+        private ByteBuffer userData;
+
+        public Assignment(List<TopicPartition> partitions, ByteBuffer userData) {
+            this.partitions = partitions;
+            this.userData = userData;
+        }
+
+        public Assignment(List<TopicPartition> partitions) {
+            this(partitions, null);
+        }
+
+        public List<TopicPartition> partitions() {
+            return partitions;
+        }
+
+        public ByteBuffer userData() {
+            return userData;
+        }
+    }
+
+    class ConsumerGroupMetadata {
+        private String groupId;
+        private int generationId;
+        private String memberId;
+        Optional<String> groupInstanceId;
+
+        public ConsumerGroupMetadata(String groupId, int generationId, String memberId, Optional<String> groupInstanceId) {
+            this.groupId = groupId;
+            this.generationId = generationId;
+            this.memberId = memberId;
+            this.groupInstanceId = groupInstanceId;
+        }
+
+        public String groupId() {
+            return groupId;
+        }
+
+        public int generationId() {
+            return generationId;
+        }
+
+        public String memberId() {
+            return memberId;
+        }
+
+        public Optional<String> groupInstanceId() {
+            return groupInstanceId;
+        }
+
+    }
+
+    final class GroupSubscription {
```

Comment:
What are the purposes of `GroupSubscription` and `GroupAssignment` classes?

Reply:
This is for possible future extensibility if we want to add new fields: with raw types (map) we'd have to change the signature of the API.